components/builder-api/src/server/authorize.rs

```diff
@@ -23,56 +24,111 @@ use crate::server::{error::{Error,
                          -> Result<originsrv::Session> {
     let session = {
         let extensions = req.extensions();
         match extensions.get::<originsrv::Session>() {
             Some(session) => {
                 let flags = FeatureFlags::from_bits(session.get_flags()).unwrap(); // unwrap Ok
                 if flags.contains(FeatureFlags::BUILD_WORKER) {
+                    debug!("authorize_session: detected allowed BUILD_WORKER");
                     return Ok(session.clone());
                 }
+                debug!("authorize_session: found session {}", session.get_id());
                 session.clone()
             }
-            None => return Err(Error::Authentication),
+            None => {
+                debug!("authorize_session: unable to get session!");
+                return Err(Error::Authentication);
+            }
         }
     };
 
     if let Some(origin) = origin_opt {
-        let mut memcache = req_state(req).memcache.borrow_mut();
-
-        match memcache.get_origin_member(origin, session.get_id()) {
-            Some(val) => {
-                trace!("Origin membership {} {} Cache Hit!",
-                       origin,
-                       session.get_id());
-                if val {
-                    return Ok(session);
-                } else {
-                    return Err(Error::Authorization);
-                }
-            }
+        let minimum_req_role = match min_role {
+            Some(r) => r,
             None => {
-                trace!("Origin membership {} {} Cache Miss!",
-                       origin,
-                       session.get_id())
+                let r = OriginMemberRole::Maintainer;
+                warn!("authorize_session: minimum role parameter not set! Assuming {}",
+                      r);
+                r
             }
-        }
-
+        };
         match check_origin_member(req, origin, session.get_id()) {
-            Ok(is_member) => {
-                memcache.set_origin_member(origin, session.get_id(), is_member);
-
-                if !is_member {
-                    return Err(Error::Authorization);
+            Ok(false) => {
+                debug!("authorize_session: account {} is not a member of origin {}",
+                       session.get_id(),
+                       origin);
+                return Err(Error::Authorization);
+            }
+            Ok(true) => {
+                match action {
+                    RoleAction::Read => {
+                        debug!("authorize_session: account {} has {} permissions in origin {}",
+                               session.get_id(),
+                               RoleAction::Read,
+                               origin);
+                        return Ok(session);
+                    }
+                    RoleAction::Write => {
+                        match check_origin_member_role(req, origin, session.get_id()) {
```

Comment:
I've added a TODO item for the use of memached optimization on `check_origin_member_role()` in https://github.com/habitat-sh/builder/pull/1378/commits/be4830739566ebad7d28d8d9157b103cf23d6a4e.

I think that's a big enough implementation delta that it merits followup as its own separate work item, distinct from the reasonably large set of changes already present in this PR.

Reply:
With regard to the optimization around combining `check_origin_member` and `check_origin_member_role`, I definitely agree with you that there is some overlap in what the two functions provide. I think that since `check_origin_member` is already a Public function in use by the server API, it would be a larger refactor lift that would need to change function return handling in multiple places.

With that in mind, and knowing that we plan to optimize `check_origin_member_role` by [leveraging memcached](https://github.com/habitat-sh/builder/issues/1384), I think it will be ok to leave these two functions distinct for now and address your suggested changes once we are informed by our optimization with memcached. 

With that in mind, and knowing that we plan to optimize `check_origin_member_role` by [leveraging memcached](https://github.com/habitat-sh/builder/issues/1384), I think it will be ok to leave these two functions distinct for now.

Reply:
@markan and I arrived at the same conclusion over a pairing session which resulted in a replacing the `check_origin_member` call with `check_origin_member_role`. This change is in https://github.com/habitat-sh/builder/pull/1378/commits/e3a581121316bba3160d4d3a3a41b380d0e28462